File: Recursion/subsequence_pattern/generate_binary_string_wiht_no_consequetive_zeros.cpp
#include <iostream>
using namespace std;
void helper(vector<string> &ans, string &ds, int n)
{
    if (ds.size() == n)
    {
        ans.push_back(ds);
        return;
    }
    ds.push_back('1');
    helper(ans, ds, n);
    ds.pop_back();
    if (ds.empty() || ds.back() == '1')
    {
        ds.push_back('0');
        helper(ans, ds, n);
        ds.pop_back();
    }
}
vector<string> validStrings(int n)
{
    vector<string> ans;
    string ds = "";
    helper(ans, ds, n);
    return ans;
}

File: Recursion/subsequence_pattern/generate_binary_string_wiht_no_consequetive_zeros.py
def helper(ans, ds, n):
    if len(ds) == n:
        ans.append(ds)
        return

    ds += '1'
    helper(ans, ds, n)
    ds = ds[:-1]

    if not ds or ds[-1] == '1':
        ds += '0'
        helper(ans, ds, n)


def valid_strings(n):
    ans = []
    helper(ans, '', n)
    return ans
